### ISSUE TITLE
JOSS review - remarks about paper

### ISSUE THREAD

**Issue by gdalle** [depasquale-lab/StateSpaceDynamics.jl#50]
Hi there!
I'm one of the reviewers for your JOSS submissions, so I'll be opening a series of issues on this repo to improve the paper, code and docs :)
Related:

- https://github.com/openjournals/joss-reviews/issues/8077

Main remarks about the paper:

- L3: Has James Fourie contributed in meaningful ways to the code? He doesn't appear on the GitHub list of contributors.
- L78: Is the code for the experiments available somewhere?
- L82: Was `dynamax` properly JIT-compiled for the experiments? I had [issues with that](https://github.com/probml/dynamax/issues/359) myself.
- L90: Is it fair to compare these packages if you only perform approximate (Laplace) EM and they perform exact EM?
- L91: Add citation for BenchmarkTools.
- L91: Benchmarking through PythonCall implies the assumption that the Julia-to-Python overhead is negligible, this assumption must be made explicit.
- L94 and L103: The benchmark plots are not readable. One option would be splitting each single plot into three plots, for instance based on the value of $n$.  Using thicker lines with different linestyles (in addition to the different markers) would also be helpful. Moreover, the value of drawing a line when there are only 3 distinct measurements is debatable: perhaps a bar plot would be easier to read?
- L94: For the first benchmark plot, I recommend replacing `julia` with `SSD` in the legend.
- L109: With proper storage, the Hessian should only require linear memory. What representation do you use?

Minor details and typos:

- L10: "is _that_ they assume"
- L16: "open-source" $\implies$ When used as adjectives, compound modifiers like these [get a hyphen](https://www.grammarly.com/blog/punctuation-capitalization/hyphen/#4). This mistake happens at lots of other places in the article, as a result I haven't listed them.
- L28: "modern neuroscientists" is a bit weird, "modern neuroscience" might be better
- L31: "precludes model _inferenece_"
- L35: "computationally _inefficeint_"
- L38: "discrete and continuous _variableS_"
- L143: There is a missing DOI in your bibliography, as spotted by the editorial bot

**Comment by gdalle** [comment]
Thanks for addressing these remarks. Here are my remaining concerns:

**Benchmarks**

- L86: Unlike Julia, JAX does not automatically get JIT-compiled with a warm-up iteration. Instead, you have to explicitly call `jax.jit` or annotate your function with the macro `@jit` for it to work (see <https://docs.jax.dev/en/latest/jit-compilation.html>).
- L99 and L108: It doesn't make a lot of sense to represent benchmarks of various sizes on the same scale. What does the bar height mean in this case? The average is not a meaningful quantity. In addition, the profusion of symbols makes the plots hard to read. I think the most principled visualization might be a table of 4 $\times$ 4 plots, each comparing the packages on instances of a given state dimension $n$ and observation dimension $m$, while varying $T$ from 100 to 1000
- L121: In your updated benchmarks, HiddenMarkovModels.jl consistently performs _worst_ and not _best_ (although I suspect that might be a benchmarking artifact, will look into it).

**Typos**

- L33: "neural activity followS"
- L56: "are block-tridiagonal"
- L99 and L108: I assume the plot titles should be $T = 100$ and not $n = 100$?

**Reproducibility**

- Code in the `benchmarking` folder should come with a `README.md`. Ideally, the experiments should be available in runnable files instead of notebooks.

**Comment by rsenne** [comment]
> Reproducibility

Similar to my comment in #51, the code you used was not up to date. This was our old benchmarking suite. I think you may not have seen my updated comment:

> Our benchmarks are still inside the benchmarking folder, but the overall structure has changed. You'll need to dev the SSD package into the benchmarking project and install the dependencies. From here you should be able to run run_benchmark.jl and run_benchmark_glmhmm.jl

This however is my fault for not having the proper branch set in the JOSS review issue. So my apologies for this! I will address the other comments and also include a properly formatted README.md for reproducibility. Thank you for all of the suggestions!

**Comment by gdalle** [comment]
Okay I profiled your use of HiddenMarkovModels.jl with `@profview` and most of the time is spent in external functions that are written suboptimally. Here are some tips to speed things up (which are also rather general):

https://github.com/depasquale-lab/StateSpaceDynamics.jl/blob/ccb0cf193875078a34c7931a7619d51dad1ca352/benchmarking/Controlled_BernoulliHMM.jl#L10

You often compute `p = logistic(z)` only to use `log(p)` afterwards. There is a faster implementation for that:

```julia
using LogExpFunctions
log_logistic(z) = -log1pexp(-z)
```

A similar shortcut exists for computing `log(1 - p)` based directly on `z`.

https://github.com/depasquale-lab/StateSpaceDynamics.jl/blob/ccb0cf193875078a34c7931a7619d51dad1ca352/benchmarking/Controlled_BernoulliHMM.jl#L39

This is much slower due to splatting a large number of arguments (with `control...`) than either of the following equivalents:

```julia
X = reduce(hcat, control)'
X = stack(control)'
```

https://github.com/depasquale-lab/StateSpaceDynamics.jl/blob/ccb0cf193875078a34c7931a7619d51dad1ca352/benchmarking/Controlled_BernoulliHMM.jl#L42

This needlessly materializes a large vector before summing it: an explicit for loop would be faster:

```julia
val = zero(Base.promote_eltype(weights, obs_seq, logits))
for i in eachindex(weights, obs_seq, logits)
    p_i = logistic(logits[i])  # to optimize, see above
    o_i = obs_seq[i]
    val -= weights[i] * (o_i * log(p_i) + (1 - o_i) * log(1 - p_i))
end
```

https://github.com/depasquale-lab/StateSpaceDynamics.jl/blob/ccb0cf193875078a34c7931a7619d51dad1ca352/benchmarking/Controlled_BernoulliHMM.jl#L58

Since the goal is to perform computations in-place, the other out-of-place computations could be rerouted to buffers instead.
For instance, you could pre-allocate `logits` before the optimization hot loop, and then in each iteration do
```julia
mul!(logits_preallocated, X, dists_coeffs)
```

https://github.com/depasquale-lab/StateSpaceDynamics.jl/blob/ccb0cf193875078a34c7931a7619d51dad1ca352/benchmarking/Controlled_BernoulliHMM.jl#L93-L94

There are a lot of common intermediate quantities between the objective and its gradients. To leverage that property, it would make sense to use the single-function interface of Optim.jl, which is [documented in NLSolversBase.jl](https://github.com/JuliaNLSolvers/NLSolversBase.jl?tab=readme-ov-file#special-single-function-interface).

**Comment by ZachLoschin** [comment]
Thank you for the useful performance tips @gdalle! I will look into this, but I am leaning towards option 2. To @gdalle's point, spending lots of time in emission model optimization introduces lots of places for discrepancies and unfair comparisons. Benchmarking a standard HMM with Gaussian emissions limits this and shifts focus back onto the fundamental HMM algorithms instead.

**Comment by gdalle** [comment]
I understand that I'm giving you a lot of headaches, perhaps more than the average JOSS review, because I'm close to the topic and I want to make sure that every package is given a fair shot. Honestly, I'd be fine with both options. For option 1, using the same optimization algorithm with the same parameters for every package (either Adam or BFGS) would be enough, provided that the implementation of the emission model is not crazy slow. Alternately, getting rid of controls is a much easier solution, and it might end up being fairer. Your call.

**Comment by gdalle** [comment]
All good now, thanks
